File: Leetcode/Trapping_Rain_water.java
// Problem Link: https://leetcode.com/problems/trapping-rain-water/


class Solution {
    public int trap(int[] a) {
        int n = a.length;
        
        if(n==0) return 0;
        
        int[] lMax = new int[n];
        int[] rMax = new int[n];
        
        lMax[0] = a[0];
        rMax[n-1] = a[n-1];
        for(int i=1;i<n;i++)
            lMax[i] = Math.max(a[i],lMax[i-1]);
        
        for(int i=n-2;i>=0;i--)
            rMax[i] = Math.max(a[i],rMax[i+1]);
        
        for(int x : lMax)
            System.out.print(x+" ");
        
        System.out.println();
    
        for(int x : rMax)
            System.out.print(x+" ");
         
        int res = 0;
        for(int i=0;i<n;i++){
            res += Math.min(lMax[i],rMax[i]) - a[i];
        }
        
        return res;
    }
}

File: Track/DSA_A_to_Z/Stacks_Queues/NGE_NLE/Trapping_Rain_Water.py
class Solution:
    def trap(self, arr: List[int]) -> int:
        n = len(arr)
        left_high = [0]*n
        right_high = [0]*n
        
        left_high[0] = arr[0]
        right_high[-1] = arr[-1]
        
        # Highest bar on left
        for i in range(1,n):
            left_high[i] = max(arr[i], left_high[i-1])
            
        # Highest bar on right
        for i in reversed(range(n-1)):
            right_high[i] = max(arr[i], right_high[i+1])
            
        res = 0
        # Max water each step can hold
        for i in range(n):
            res += min(left_high[i], right_high[i]) - arr[i]
        
    
        return res
        